src/inmanta/env.py

```diff
@@ -86,6 +88,42 @@ def get_packages_in_working_set(cls) -> Dict[str, version.Version]:
     def rebuild_working_set(cls) -> None:
         pkg_resources.working_set = pkg_resources.WorkingSet._build_master()
 
+    @classmethod
+    def get_dependency_tree(cls, dists: abc.Iterable[str]) -> abc.Set[str]:
+        """
+        Returns the full set of all dependencies (both direct and transitive) for the given distributions. Includes the
+        distributions themselves.
+        If one of the distributions or its dependencies is not installed, it is still included in the set but its dependencies
+        are not.
+
+        :param dists: The keys for the distributions to get the dependency tree for.
+        """
+        # create dict for O(1) lookup
+        installed_distributions: abc.Mapping[str, Distribution] = {
+            dist_info.key: dist_info for dist_info in pkg_resources.working_set
+        }
+
+        def _get_tree_recursive(dists: abc.Iterable[str], acc: abc.Set[str] = frozenset()) -> abc.Set[str]:
+            """
+            :param acc: Accumulator for requirements that have already been recursed on.
+            """
+            return reduce(_get_tree_recursive_single, dists, acc)
+
+        def _get_tree_recursive_single(acc: abc.Set[str], dist: str) -> abc.Set[str]:
+            if dist in acc:
+                return acc
+
+            if dist not in installed_distributions:
+                return acc | {dist}
+
+            # recurse on direct dependencies if they have not yet been recursed on in a previous iteration
```

Comment:
```suggestion
            # recurse on direct dependencies
```
The second part is no longer relevant: we recurse anyway, it just happens to be a noop if it has already been processed. My bad.